torch/distributed/_tensor/ops/view_ops.py

```diff
@@ -618,18 +614,30 @@ def reshape_prop(op_schema: OpSchema) -> OutputSharding:
         if shard_out is not None:
             # no reshard needed
             output_dtensor_spec = DTensorSpec(mesh=mesh, placements=tuple(shard_out))
-            local_out_shape = compute_local_shape(
-                list(global_out_shape), mesh, shard_out
-            )
 
             # We only need the local shape to lower the call into the local op
             args = op_schema.args_schema
             shape_argnum = spec.shape_argnum
             if shape_argnum is not None:
-                op_schema.args_schema = (
-                    args[:shape_argnum]
+                # compute the local shape from the global shape, then return
+                # a resharding even if we don't really reshard, the only reason
+                # for this type of resharding is to lower the global shape to
+                # local shape
+                local_out_shape = compute_local_shape(
+                    list(global_out_shape), mesh, shard_out
+                )
+
+                suggested_schema = OpSchema(
+                    func_schema=op_schema.func_schema,
+                    args_schema=args[:shape_argnum]
                     + (tuple(local_out_shape),)
-                    + args[shape_argnum + 1 :]
+                    + args[shape_argnum + 1 :],
+                    kwargs_schema=op_schema.kwargs_schema,
+                )
+                return OutputSharding(
+                    output_spec=output_dtensor_spec,
+                    schema_suggestions=[suggested_schema],
+                    needs_redistribute=True,
```

Comment:
Nice! It would be interesting to see if we still have any cache miss on both MLP mini example and transformer model. For example, detach and index op I observed in the past which also have op_schema changing.